docs/design/variadics.md

```diff
@@ -0,0 +1,608 @@
+  ...{ result.Append(each param.ToString()); }
+  return result;
+}
+```
+
+```carbon
+// Returns the minimum of its arguments, which must all have the same type T.
+//
+// Note that this implementation is not recursive. We split the parameters into
+// first and rest in order to forbid calling `Min` with no arguments.
+fn Min[T:! Comparable & Value](first: T, ..., each next: T) -> T {
+  var result: T = first;
+  ...{
+    if (each next < result) {
+      result = next;
+    }
+  }
+  return result;
+}
+```
+
+```carbon
+// Invokes f, with the tuple `args` as its arguments.
+fn Apply[each T:! type, F:! CallableWith(..., each T)](f: F, args: (..., each T)) -> auto {
+  return f(..., expand args);
+}
+```
+
+```carbon
+// Toy example of mixing packs and values in a single parameter list.
+// Takes an i64, any number of f64s, and then another i64.
+fn MiddleVariadic(first: i64, ..., each middle: f64, last: i64);
+```
+
+```carbon
+// Toy example of using the result of variadic type deduction.
+fn TupleConcat[each T1: type, each T2: type](
+    t1: (..., each T1), t2: (..., each T2)) -> (..., each T1, ..., each T2) {
+  return (..., expand t1, ..., expand t2);
+}
+```
+
+## Expression and statement semantics
+
+In all of the following, N is the arity of the pack expansion being discussed,
+and `$I` is a notional variable representing the pack index. These semantics are
+implemented at monomorphization time, so the value of N is a known integer
+constant. Although the value of `$I` can vary during execution, it is
+nevertheless treated as a constant; for example, it can be used to index a
+tuple.
+
+A statement of the form "`...{` _block-contents_ `}`" is evaluated by executing
+_block-contents_ N times, with `$I` ranging from 0 to N - 1.
+
+An expression of the form "`...and` _operand_" is evaluated as follows: a
+notional `bool` variable `$R` is initialized to `true`, and then "`$R = $R and`
+_operand_" is executed up to N times, with `$I` ranging from 0 to N - 1. If at
+any point `$R` becomes false, this iteration is terminated early. The final
+value of `$R` is the value of the expression.
+
+An expression of the form "`...or` _operand_" is evaluated the same way, but
+with `or` in place of `and`, and `true` and `false` transposed.
+
+An expression of the form "`...,` _operand_" evaluates to a sequence of N
+values, where the k'th value is the value of _operand_ where `$I` is equal to
+k - 1.
+
+An expression of the form "`expand` _argument_" evaluates to "_argument_
+`[$I]`".
+
+An expression of the form "`each` _identifier_", where _identifier_ names a
+variadic binding, evaluates to the `$I`th value that it was bound to (indexed
+from zero).
+
+## Typechecking expressions and statements
+
+### Generalized tuple types
+
+The `...,` operator lets us form tuples out of sequences whose size is not known
+during typechecking. For example, in this code:
+
+```carbon
+fn F[each T:! type](x: (..., each i32), ..., each y: Optional(each T)) {
```

Comment:
`each i32` doesn't seem to be valid per the earlier rules, as `i32` is not a variadic binding nor a binding pattern. Should this be:

```suggestion
fn F[each T:! type](x: (..., i32), ..., each y: Optional(each T)) {
```

... where the pack expansion in the type of `x` doesn't contain any expansion arguments? While that would normally be problematic, I think it's OK in the type of a pattern because we can deduce the arity based on usage.

Reply:
Isn't the only thing being deduced is the arity of the tuple, and that fact is being marked by `...`?

Reply:
`...` is also used in a lot of non-pattern contexts where deduction doesn't happen, unlike `auto` and `;]`, which dilutes its value as a marker of deduction. Even in patterns it's not a fully reliable signal -- e.g. there's no arity deduction in a pattern like `x: (... expand (i32, i32))` (which is silly but valid).